File: src/shell_cluster/tunnel/discovery.py
"""Peer discovery via tunnel backend."""

from __future__ import annotations

import asyncio
import logging

from typing import Awaitable, Callable

from shell_cluster.models import Peer, PeerStatus
from shell_cluster.tunnel.base import TunnelBackend, parse_node_name

log = logging.getLogger(__name__)

OnPeersChanged = Callable[[list[Peer]], Awaitable[None]]


class PeerDiscovery:
    """Discovers peers by listing tunnels with a shared label."""

    def __init__(
        self,
        backend: TunnelBackend,
        label: str,
        own_tunnel_id: str,
        interval: int = 30,
        on_peers_changed: OnPeersChanged | None = None,
    ):
        self._backend = backend
        self._label = label
        self._own_tunnel_id = own_tunnel_id
        self._interval = interval
        self._on_peers_changed = on_peers_changed
        self._peers: dict[str, Peer] = {}
        self._running = False

    @property
    def peers(self) -> dict[str, Peer]:
        return dict(self._peers)

    async def refresh(self) -> list[Peer]:
        """Refresh the peer list from the tunnel backend."""
        try:
            tunnels = await asyncio.wait_for(
                self._backend.list_tunnels(self._label), timeout=10.0
            )
        except asyncio.TimeoutError:
            log.warning("Discovery refresh timed out")
            return list(self._peers.values())
        except Exception as e:
            log.warning("Discovery refresh failed: %s", e)
            return list(self._peers.values())

        seen: set[str] = set()
        for t in tunnels:
            # Skip tunnels that are not actively hosted
            if not t.hosting:
                continue
            seen.add(t.tunnel_id)

            existing = self._peers.get(t.tunnel_id)
            if existing and existing.status == PeerStatus.ONLINE:
                # Already online — check if port changed (peer restarted between cycles)
                if t.port and t.port != existing.port:
                    log.info("Peer %s port changed %d -> %d (list_tunnels), refreshing",
                             existing.name, existing.port, t.port)
                    remote_port, forwarding_uri = await self._backend.get_port_and_uri(t.tunnel_id)
                    if remote_port and remote_port > 0:
                        existing.port = remote_port
                        existing.forwarding_uri = forwarding_uri
            elif existing:
                # Was offline, now back online — re-query port (may have changed)
                log.info("Peer %s back online, refreshing port info", existing.name)
                remote_port, forwarding_uri = await self._backend.get_port_and_uri(t.tunnel_id)
                if remote_port and remote_port > 0:
                    existing.port = remote_port
                    existing.forwarding_uri = forwarding_uri
                    existing.status = PeerStatus.ONLINE
                else:
                    log.warning("Peer %s returned invalid port %s, keeping offline", existing.name, remote_port)
            else:
                # New peer discovered — get port info
                remote_port, forwarding_uri = await self._backend.get_port_and_uri(t.tunnel_id)
                if not remote_port or remote_port <= 0:
                    log.warning("New peer %s has invalid port %s, skipping", t.tunnel_id, remote_port)
                    continue
                name = parse_node_name(t.tunnel_id)
                peer = Peer(
                    name=name,
                    tunnel_id=t.tunnel_id,
                    port=remote_port,
                    forwarding_uri=forwarding_uri,
                    status=PeerStatus.ONLINE,
                )
                self._peers[t.tunnel_id] = peer
                log.info("Discovered peer: %s (%s)", name, t.tunnel_id)

        # Mark unseen peers as offline
        for tid in list(self._peers):
            if tid not in seen:
                self._peers[tid].status = PeerStatus.OFFLINE

        return list(self._peers.values())

    async def run_loop(self) -> None:
        """Run discovery in a loop, calling on_peers_changed when list changes."""
        self._running = True
        while self._running:
            peers = await self.refresh()
            if self._on_peers_changed:
                await self._on_peers_changed(peers)
            await asyncio.sleep(self._interval)

    def stop(self) -> None:
        self._running = False

File: server/src/tunnel/discovery.js
/**
 * Peer discovery via tunnel backend (Node.js port).
 */

'use strict';

const { parseNodeName } = require('./base');

const PEER_STATUS = {
  ONLINE: 'online',
  OFFLINE: 'offline',
  CONNECTING: 'connecting',
};

class PeerDiscovery {
  /**
   * @param {object} opts
   * @param {object} opts.backend - tunnel backend instance
   * @param {string} opts.label
   * @param {string} opts.ownTunnelId
   * @param {number} [opts.interval=300]
   * @param {function(Array):Promise} [opts.onPeersChanged]
   */
  constructor(opts) {
    this._backend = opts.backend;
    this._label = opts.label;
    this._ownTunnelId = opts.ownTunnelId;
    this._interval = opts.interval || 300;
    this._onPeersChanged = opts.onPeersChanged || null;
    this._peers = new Map(); // tunnelId -> peer object
    this._running = false;
    this._timer = null;
  }

  get peers() {
    return new Map(this._peers);
  }

  /**
   * Refresh the peer list from the tunnel backend.
   * @returns {Promise<Array>}
   */
  async refresh() {
    let tunnels;
    try {
      tunnels = await Promise.race([
        this._backend.listTunnels(this._label),
        new Promise((_, reject) => setTimeout(() => reject(new Error('timeout')), 10000)),
      ]);
    } catch (e) {
      console.warn(`[Discovery] Refresh failed: ${e.message}`);
      return Array.from(this._peers.values());
    }

    const seen = new Set();

    for (const t of tunnels) {
      if (!t.hosting) continue;
      seen.add(t.tunnelId);

      const existing = this._peers.get(t.tunnelId);
      if (existing && existing.status === PEER_STATUS.ONLINE) {
        // Check if port changed
        if (t.port && t.port !== existing.port) {
          console.log(`[Discovery] Peer ${existing.name} port changed ${existing.port} -> ${t.port}`);
          const result = await this._backend.getPortAndUri(t.tunnelId);
          if (result.port > 0) {
            existing.port = result.port;
            existing.forwardingUri = result.uri;
          }
        }
      } else if (existing) {
        // Was offline, now back online
        console.log(`[Discovery] Peer ${existing.name} back online, refreshing port`);
        const result = await this._backend.getPortAndUri(t.tunnelId);
        if (result.port > 0) {
          existing.port = result.port;
          existing.forwardingUri = result.uri;
          existing.status = PEER_STATUS.ONLINE;
        }
      } else {
        // New peer discovered
        const result = await this._backend.getPortAndUri(t.tunnelId);
        if (!result.port || result.port <= 0) {
          console.warn(`[Discovery] New peer ${t.tunnelId} has invalid port, skipping`);
          continue;
        }
        const name = parseNodeName(t.tunnelId);
        const peer = {
          name,
          tunnelId: t.tunnelId,
          port: result.port,
          forwardingUri: result.uri,
          status: PEER_STATUS.ONLINE,
        };
        this._peers.set(t.tunnelId, peer);
        console.log(`[Discovery] Discovered peer: ${name} (${t.tunnelId})`);
      }
    }

    // Mark unseen peers as offline
    for (const [tid, peer] of this._peers) {
      if (!seen.has(tid)) {
        peer.status = PEER_STATUS.OFFLINE;
      }
    }

    return Array.from(this._peers.values());
  }

  /**
   * Run discovery in a loop.
   * @param {object} [opts]
   * @param {boolean} [opts.skipFirst=false]
   */
  async runLoop(opts = {}) {
    this._running = true;
    let first = true;

    while (this._running) {
      if (first && opts.skipFirst) {
        first = false;
      } else {
        const peers = await this.refresh();
        if (this._onPeersChanged) {
          await this._onPeersChanged(peers);
        }
      }

      // Sleep with cancellation support
      await new Promise((resolve) => {
        this._timer = setTimeout(resolve, this._interval * 1000);
      });
    }
  }

  stop() {
    this._running = false;
    if (this._timer) {
      clearTimeout(this._timer);
      this._timer = null;
    }
  }
}

module.exports = { PeerDiscovery, PEER_STATUS };
